lib/rules/array-bracket-newline.js

```diff
@@ -0,0 +1,247 @@
+        */
+        function reportRequiredBeginningBreak(node, token) {
+            context.report({
+                node,
+                loc: token.loc.start,
+                message: "A break is required after '{{tokenValue}}'.",
+                data: {
+                    tokenValue: token.value
+                },
+                fix(fixer) {
+                    return fixer.insertTextAfter(token, "\n");
+                }
+            });
+        }
+
+        /**
+        * Reports that there should be a break before the last token
+        * @param {ASTNode} node - The node to report in the event of an error.
+        * @param {Token} token - The token to use for the report.
+        * @returns {void}
+        */
+        function reportRequiredEndingBreak(node, token) {
+            context.report({
+                node,
+                loc: token.loc.start,
+                message: "A break is required before '{{tokenValue}}'.",
+                data: {
+                    tokenValue: token.value
+                },
+                fix(fixer) {
+                    return fixer.insertTextBefore(token, "\n");
+                }
+            });
+        }
+
+        /**
+         * Reports a given node if it violated this rule.
+         *
+         * @param {ASTNode} node - A node to check. This is an ObjectExpression node or an ObjectPattern node.
+         * @param {{multiline: boolean, minItems: number}} options - An option object.
+         * @returns {void}
+         */
+        function check(node) {
+            const elements = node.elements;
+            const normalizedOptions = normalizeOptions(context.options[0]);
+            const options = normalizedOptions[node.type];
+            const openBracket = sourceCode.getFirstToken(node);
+            const closeBracket = sourceCode.getLastToken(node);
+            let first = sourceCode.getTokenAfter(openBracket, { includeComments: true });
+            let last = sourceCode.getTokenBefore(closeBracket, { includeComments: true });
+
+            const needsLinebreaks = (
+                elements.length >= options.minItems ||
+                (
+                    options.multiline &&
+                    elements.length > 0 &&
+                    first.loc.start.line !== last.loc.end.line
+                )
+            );
+
+            /*
+             * Use tokens or comments to check multiline or not.
+             * But use only tokens to check whether line breaks are needed.
+             * This allows:
+             *     var arr = [ // eslint-disable-line foo
+             *         'a'
+             *     ]
+             */
+            first = sourceCode.getTokenAfter(openBracket);
+            last = sourceCode.getTokenBefore(closeBracket);
+
+            if (needsLinebreaks) {
+                if (astUtils.isTokenOnSameLine(openBracket, first)) {
+                    reportRequiredBeginningBreak(node, sourceCode.getFirstToken(openBracket));
```

Comment:
Suppose the code looks like this:

```js
[
  foo,
  bar
]
```

Right now, the `reportNoBeginningBreak` function gets called with the `foo` token, and then there is some logic in the function to get the `[` token. However, I think it would be better if `reportNoBeginningBreak` was called directly with the `[` token.

Similarly, right now `reportNoEndingBreak` gets called with the `bar` token, but I think it would be better if it was called directly with the `]` token.

Does that make more sense? Let me know if I should explain it further.